Desktop: #5867 Add support for "categories" manifest field

mak2002 opened the pull request:
- Issue addressed #5867 
## Overview
Finding the right plugin can be difficult as there are more than 80 plugins now. For implementing displaying plugins by category, we can divide this into 2 subtasks.
1. We will first need to allow plugin devs to set the plugins' categories from a pre-defined set of categories as discussed in the above issue.
2. Add ability to filter plugins by category in Joplin app.

This PR will be focusing on 1. 
## Solution
To achieve this, I have done the following:
- Added a **categories** field in '*Plugin Manifest*' interface as well as *manifest.json* template. In my opinion this field should **not be** **required** because currently Joplin is fetching **Manifest.json** from 'joplin/plugins' repo and category field is not present in any manifest.
	- Also for reference, VS code's extension manifest don't make category field required ([VS Code Extension Manifest](https://code.visualstudio.com/api/references/extension-manifest)).
-  Added function in webpack.config.js(generator-joplin) to validate categories given by plugin dev.
-  I have taken the categories from this issue disscused in this [Issue #5867](https://github.com/laurent22/joplin/issues/5867). I have excluded some categories as some can go under another umbrella category. For ex: email (excluded) -> integration.
	-   Categories: `"administration", "developer", "editor", "export", "import", "integration", "interface", "linking",
								"markdown enhancements", "note management", "rendering", "tagging", "theme", "todo", "search", "other"`
	-	Currently, I have limited the number of categories a plugin can have to 3 as I could not find a plugin that needed more than 3 categories to describe general function of the plugin but I am open to change.
	- I am ok with changing the category names if you want.
## Testing
I have manually tested the webpack with the following cases:
- With repeated categories
- with invalid category
- with "" as a category
- category with accented character

laurent22 commented:
I'm not a fan of the category names. "Others" for sure should go. And the rest I don't know as it feels low level or vague. You have to imagine someone browsing by categories at a high level. For example "rendering" that's for us programmers but the average user might not know what it means. Perhaps "visualisation"? 

Getting the categories right is the most important part of the task because once they are set we'll try hard not to change them, so a bit more thoughts need to go into them.

My suggestion would be to look at other app plugin system categories - don't check just one app, check Firefox, Chrome, VSCode, Obsidian and another app you can think of, and look at what categories they use. Make a list of all their categories. Maybe we can borrow some of them, or at least it will give us an idea of what a "category" should be.

Also should we really allow multiple categories considering we already have keywords?

mak2002 commented:
Thanks for the suggestions. I will look into those and update you soon.

tessus commented:
> Also should we really allow multiple categories considering we already have keywords?

Yes.

The rest I agree with you.

tessus commented:
Let me explain my short and concise `Yes`.

> should we really allow multiple categories considering we already have keywords?

Why would we need a category then in the first place? We already have keywords.

I am not saying we should be able to search for multiple categories or a allow selecting multiple categories, but a plugin can still belong to different categories.

Take a movie on Netflix. The same movie might show up in the categories `Sci-Fi` , `Fantasy`, and `Action`.
Btw, the Apps organization in Nextcloud works the same way. An app can show up in differegnt categories.

mak2002 commented:
- As suggested by @laurent22  I looked up a couple of other app plugin system categories, For reference I am adding all of them here, Also I have highlighted categories in the reference that could be useful for us.
- And finally I have condensed them to a few that will be useful for us. 

#### 1\. Chrome Extension Categories:

- _**Recommended for you**_, **_Accessibility_**, Blogging, **_Developer_ tools**, Fun, News and weather, photos, **_productivity_**, **_search tools_**, shopping, social and communication, sports, **_themes_**


#### 2\. Firefox Addons Categories:

- Alerts & Updates, **Appearance**, Bookmarks, Download Management, Feeds, News & Blogging, Games & Entertainment
    Language Support, Photos, Music & Videos, Privacy & Security, **Search Tools** , Shopping, Social & Communication, Tabs, **Web Development**

#### 3.Notion Tools:(Curated list website built by an individual)

- **Templates**, **Personalization**, Apps & Plugins, **Website Building**, Resources, Communication, **Integrations**, Content Creators

#### 4\. Obsidian Categories from [Forum Post](https://forum.obsidian.md/t/plugin-categorisation/13565/8)

- View, Edit, Integration, Task, Shortcut, **Link**, **Style**, **Automation**, Convert, **Search**, Refactor, Render, Navigation, **File**, History, **Template**, Editor, Miscellaneous, PDF, **Statistics**, Chart, Chinese, Game, Image, Mindmap, Security, Vim, Attachment, **Backup**, Date, **Math**, Mobile, Video/Audio, Calendar, Financial, Japanese, Journal, Workspace

#### 5\. Nextcloud Apps Categories and many are similar to ownCloud

**Customization**, Dashboard, **Files**, Games, Integration, Monitoring, Multimedia, Office and text, Organization, **Search**, Security, Social and communication, Tools, Flow

#### My Own Suggstions:

- Education, Files and Backup , Editor Enhancer

#### I have condensed some categories for our use case as following:

- Recommended, Appearance, Developer tools, Productivity, Themes, Templates, Integrations, Todos, Note Management, Linking, Education,Tags, Files and Backup, Editor Enhancer, Personalization/Customization(one of these), Import/Export(both as one)



> should we really allow multiple categories considering we already have keywords?

I also agree with tessus that we should allow multiple categories. For example, 'Untagged Plugin' allows to to find notes and todos without a tag so this can also come under both 'Tags' and 'Note Management' category.

mak2002 commented:
hey @tessus ,Can you please take a look at these categories and give your suggestions. Then after finalizing we can ask Laurent for his approval. 

> * `Recommended, Appearance, Developer tools, Productivity, Themes, Templates, Integrations, Todos, Note Management, Linking, Education, Search Tools, Tags, Files and Backup, Editor Enhancer, Personalization/Customization(one of these), Import/Export(both as one) `
> 
> >

mak2002 commented:
Sorry if I am missing something, Can any of the moderators tell me what should I do further?

mak2002 commented:
> Same with 'Education'

Currently there are 2 plugins in this category **Math Mode and turnToChart**. Maybe this category is too specific and hence not needed.

tessus commented:
I am not sure what to do with `Recommended`. Laurent said that devs should be able to set the categories. But if we used a `Recommended` category, it must not be set by devs, but only by the Joplin team. Thus I think you mightbe able to remove that category from the list. Or at least from the validity check.

`Editor` is definitely better than `Editor Enhancer`.

Also, you should settle on one Personalization/Customization. I think `Customisation` (we use British spelling) is more common that Personalisation, but that's just my opinion.

laurent22 commented:
I agree with tessus comment. Also what is "Integrations"? Otherwise "Search tools" => "Search".

And I think there's an overlap between "Files and Backup" and "Import/Export". Do we even need a "Backup" category as it's unlikely there will be more than the Backup plugin, and "Files" is bit vague. Most "Files" will also be "Import/Export".

mak2002 commented:
Response to Tessus comment:
> But if we used a `Recommended` category, it must not be set by devs

I missed it. You are right I will remove it from validity check. But in my opinion we should keep it as a category set by Joplin team.


> I think `Customisation` (we use British spelling) is more common that Personalisation

Yep, I also liked `Customisation`.
___
Response to Laurent's comment:
> what is "Integrations"? Otherwise "Search tools" => "Search".

This category can include plugins that integrates third party apps with Joplin. For example:
- 'Jira Issue' retrieves Atlassian Jira issues information using their api in order to track the status of them from your Joplin notes.
- Another example is 'Joplin Anki Sync' it is Bidirectional sync between Joplin and Anki.

> Do we even need a "Backup" category

I also think there wont be many Backup plugins. So this category will be too specific and wont be needed.

> Most "Files" will also be "Import/Export".

I agree. I will remove 'Files and Backup' then.

Thanks for the feedback.

So the updated categories are:
`Recommended(set by Joplin team), Appearance, Developer tools, Productivity, Themes, Templates, Integrations, Note Management, Search, Tags, Editor, Customization, Import/Export`

mak2002 commented:
Hey Laurent, Are you ok with "Integrations" category? If so then I will add a commit with updated ones.

Daeraxa commented:
> I think all themes could go under the "Appearance" category, so we could remove the "Themes" category. What do you guys think?

I do think there is (or should be) a distinction between theme and appearance. I would think a theme would be an entire theme (like the macOS or white themes already published) but "appearance" could cover other items like the line numbers or "markdown table: colorize" plugins.

If there was a future decision to move theme plugins under the actual `appearance` configuration page then pre-categorising them ahead of time might make sense (i.e. if you see a 'theme' category it is no longer eligible to show in the plugins search but shows in a potential themes search instead).

mak2002 commented:
> isn't "Customisation" too broad?

- I think this probably wont be primary or the only category to describe a plugin but it may act as a secondary category. As far as which plugins may go under this category following are some examples: `Menu items, Shortcuts, Toolbar icons`, `Favorites` (Save any notebook, note, to-do, tag favorite in an extra panel view), `Folding in Code Mirror Editor`(Allows to fold parts of markdown text displayed in editor).
- But I agree with you that plugin devs may think that almost all plugins would fall under this category. But as you said I can find new name for this.

> How about we replace "Import/Export" by just "Files"?

Sure, This will simplify things.

> all themes could go under the "Appearance" category

- I have same opinion as Daeraxa (stated above). There are separate plugins especially for theme like `GitHub Theme`, `White Theme`  
- I have been thinking of adding a check to see if there is already 'theme' category then plugin dev shouldn't be able to add 'appearance' category to avoid repetition. Do you think this will be good idea?

mak2002 commented:
Also can I open another PR implementing UI for this as this will speed up the things? I thought this will be good time to do this because this is still pre GSOC.

laurent22 commented:
We'd need to know what UI needs to be implemented first, as the original issue doesn't have much details about this.

laurent22 commented:
> I have same opinion as Daeraxa (stated above). There are separate plugins especially for theme like GitHub Theme, White Theme

Ok that makes sense, let's keep both "appearance" and themes" then.

> plugins may go under this category following are some examples: Menu items, Shortcuts, Toolbar icons, Favorites (Save any notebook, note, to-do, tag favorite in an extra panel view), Folding in Code Mirror Editor(Allows to fold parts of markdown text displayed in editor).

Still not sure about this "Customisation" category. "Folding in Code Mirror Editor" can go in the "editor" category, and "Favorites" could go in "note management".

I indeed don't know about "Menu items", "Shortcuts" and Toolbar icons" as they are relatively generic. But on the other hand if "customisation" becomes the same as "others" we may as well have an "others" category as that would be clearer what it is. Ideally we should find some category though, any suggestions?

Other than this category, I think the rest is fine now.

mak2002 commented:
> We'd need to know what UI needs to be implemented first, as the original issue doesn't have much details about this.

I have implemented the following for demo purpose. I will use more consistent styling for drop-down menu.

https://user-images.githubusercontent.com/77744862/156114015-dcc41ae6-ed29-4181-94de-81aaa103d1db.mp4

What do you guys think?

laurent22 commented:
Hmm, I'm not sure. As always we need to discuss what needs to be done before jumping straight to coding. The way I see it categories are good for browsing plugins from a website (that currently doesn't exist), or possibly from the app itself but how it should be done is not clear. What other apps are doing?

Daeraxa commented:
VSCode definitely has one of the more comprehensive extensions system in regards to categorisation and filtering:
![image](https://user-images.githubusercontent.com/58074586/156160122-e3dc2543-7593-434d-bcb2-3819af8aa636.png)

Obsidian and Atom both have more or less the same as what Joplin offers currently it seems.

mak2002 commented:
> we may as well have an "others" category

Yeah, I think you are right. I think we need a `others` category otherwise in some cases we may be forcing plugins dev to include some irrelevant category. 

Current Categories:
` Appearance, Developer tools, Productivity, Themes, Integrations, Note Management, Search, Tags, Editor, Files, Others,`
> As always we need to discuss what needs to be done

Ok, I will make new topic for this in forum.

I have been thinking of adding a check to see if there is `Themes` category then plugin dev shouldn't be able to add `Appearance` category to avoid repetition. Do you think this will be good idea?

laurent22 commented:
> Yeah, I think you are right. I think we need a `others` category otherwise in some cases we may be forcing plugins dev to include some irrelevant category.

I think you missed the part "Ideally we should find some category though, any suggestions?". Don't you think an "others" category is not useful since that field is optional? If the developer doesn't have a category then they can simply not put any. But our task is to find some that cover all cases. It seems the last plugins we didn't categorise are "Menu items", "Shortcuts" and Toolbar icons".

> I have been thinking of adding a check to see if there is Themes category then plugin dev shouldn't be able to add Appearance category to avoid repetition. Do you think this will be good idea?

We won't have any complex logic here for now. Any such logic should be on the front end anyway, for example in the way we filter the plugins by taking into account (or not) the categories that the author specified.

mak2002 commented:
> I think you missed the part "Ideally we should find some category though, any suggestions?". Don't you think an "others" category is not useful since that field is optional?

Ohh sorry. I thought you were suggesting "Others" as a category. Also we can treat plugins without any categories as "Others" in the app so there would be no need add it specifically.

> It seems the last plugins we didn't categorise are "Menu items", "Shortcuts" and Toolbar icons".

I will soon update you for this one.

mak2002 commented:
> I indeed don't know about \"Menu items\", \"Shortcuts\" and Toolbar icons\" as they are relatively generic

Do you think we must add a new category specifically for this plugin alone? As this plugin deals with different kind of things.

laurent22 commented:
Oh sorry I thought that was three different plugins. Hmm, then maybe let's not do anything about it for now. Perhaps we can think of a new category later on that can include this plugins and others.

mak2002 commented:
Thanks everybody for the feedback on this one!
Categories: 
`Appearance, Developer tools, Productivity, Themes, Integrations, Note Management, Search, Tags, Editor, Files`
Just to confirm do you agree on having 3 categories limit?

laurent22 commented:
> Just to confirm do you agree on having 3 categories limit?

It doesn't really matter at this point as we don't put restrictions on the manifest file, since we don't control it anyway (so even if there's validation, it's easy to bypass it). When processing these files however we may only consider the first 2 or 3 listed categories, drop invalid ones, and in some cases consider only the first one.

We could document this, for example by suggesting the developers to list the categories in order, from the most relevant to less relevant one.

roman-r-m commented:
Are there any plugins that fit into the Developer Tools category?

mak2002 commented:
Yes, 'Get Notebook ID' and 'Conflict Resolution' can fit into this category.

laurent22 commented on packages/generator-joplin/generators/app/templates/webpack.config.js:
```diff
@@ -32,9 +32,7 @@ const packageJsonPath = `${rootDir}/package.json`;
-const allPossibleCategories = ["Recommended", "Appearance"," Developer tools", "Productivity", "Themes",
-								"Templates", "Integrations", "Todos", "Note Management", "Linking", "Education",
-								"Search Tools", "Tags", "Files and Backup", "Editor Enhancer", "Customization", "Import/Export"]
+const allPossibleCategories = ['Appearance', 'Developer tools', 'Productivity', 'Themes', 'Integrations', 'Note Management', 'Search', 'Tags', 'Editor', 'Files'];
```
Note Management => Note management

mak2002 replied:
Done :+1:

laurent22 commented:
Would it make sense to take the plugin list from [here](https://github.com/joplin/plugins) and add a category to each of them, to see if what we have is comprehensive?

mak2002 commented:
Sure! I'm on it👍

mak2002 commented:
> add a category to each of them

But strictly one category or multiple are also fine?

mak2002 commented:
> add a category to each of them

Sorry if this is silly question but should I strictly add one or multiple categories are also fine?

laurent22 commented:
I'd think one is enough, unless you feel some should have two or more.

mak2002 commented:
I added category to every plugin and added multiple categories to 13 plugins. I have given category name with number of respective plugins in brakets. 
- Appearance(9), Developer tools(4), Productivity(9), Themes(3), Integrations(12), Note Management(30), Search(3), Tags(6), Editor(24), Files(9)

Should I also add table containing each plugin with respective plugin?

laurent22 commented:
Yes if you could add put the table here in the comments that would be great. Perhaps we can split the Note Management category into something else. The last 3 you mentioned perhaps could go in Productivity.

mak2002 commented:
First column isn't showing images but works fine.

|  Homepage   | Name | Categories | Description |
| --- | --- | --- | --- |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/joplin/plugin-abc-sheet-music) | ABC Sheet Music Plugin | editor | Turns ABC text notation into sheet music |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/maxnegro/joplin-plugin-admonition) | Admonition markdown extension | note management | Enables markdownIt admonition plugin, which enables formatting notes |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/agenda/21158) | Agenda | Productivity | An agenda/calendar/schedule panel for joplin that shows all uncompleted to-dos with a due date. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/makaanneo/joplin-plugin-athena) | Athena | Files | Import files from import a folder into joplin notebook as notes mit metadata (PDF text in note comment, no OCR) |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/manuerwin/joplin-plugin-attache#readme) | Attaché | Files | Your attachment updater. Mass replacement of Joplin attachments (resources) such as resized image files, current music playlists, and any other attachments you need to one-off or regularly update within Joplin. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/joplin/plugins/blob/master/github.com/zph/joplin-plugin-auto-alarm) | Auto Alarm | Productivity | Natural language entry of alarm date/times |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/eyalag/joplin-plugin-auto-show-active-note) | Auto show active note in sidebar | note management, appearance | Auto show active note in sidebar |
| -   | Auto Tag | tags | Convert all octothorpe prefixed words into tags |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/S73ph4n) | Autolinker | note management | Adds an icon to the toolbar. When clicked, it creates a link to the note with a title matching the selected text. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/S73ph4n) | Automate Notes | developer tools, editor | Finds blocks of javascript in the current note (blocks have to start with '`javascript' and end with '`') and runs it as a function. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/insert-referencing-notes-backlinks-plugin/13632) | Automatic Backlinks to note | note management | Creates backlinks to opened note, also in automatic way |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ylc395/joplin-plugin-better-markdown-viewer) | Better Markdown Viewer | editor | Keep Cursor Synced Between Editor and MD Viewer |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/DanteCoder/JoplinBibleQuote#readme) | Bible Quote | integrations | Quote the bible direct in text editor |
| -   | BibTeX | integration | Use locally stored BibTeX files to integrate citation into Joplin |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ahaltindis/joplin-plugin-bulk-note-creator) | Bulk Note Creator | note management | Create notes as bulk |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/marc0l92/joplin-plugin-bytefield-svg#readme) | bytefield-svg | editor | Add support for bytefield-svg diagrams used to describe network protocols, memory layouts and any other binary structure. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/Mick2nd/Code-Section#readme) | Code Section | editor, appearance | Inserts a piece of code from code source |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/shantanugoel/joplin-plugin-cm-linenumbers) | CodeMirror Line Numbers | editor, appearance | Enables line numbers for CodeMirror editor |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/JackGruber/joplin-plugin-combine-notes) | Combine notes | note management | Combine one or more notes |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/S73ph4n) | Complete Link | note management | This plugin adds an icon to the toolbar. When clicked, it completes the selected text into a note title (in a markdown link). |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/plugin-conflict-resolution/19204) | Conflict Resolution | developer tools | A plugin that makes process of resolving conflicts easier. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-container-with-classes) | Container with Classes | developer tools | Plugin for creating block-level containers with classes for customization |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/create-note-from-highlighted-text/12511) | Convert Text To New Note | note management | Converts highlighted text to new one in same folder |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-copy-anchor-link) | Copy Anchor Link | note management | Add a link icon near each heading. Copy Joplin markdown link of the specific heading to clipboard when click on the icon. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/copy-markdown-link-to-active-note/14402) | Copy link to active note | note management | Adds entry to right click menu in editor to get link to active note |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/go-to-note-tag-or-notebook-via-highlighting-text-in-editor/12731) | Create and go to #tags and @notebooks | note management | Go to tag,notebook or note via links or via text |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/Mick2nd/CsvImport#readme) | Csv Import | files | Imports a Csv file or pastes a Csv text from Clipboard as table |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/S73ph4n) | Day Review | productivity | This plugin adds an icon to the toolbar. When clicked, it makes a review of the day's created/updated/completed notes and todos. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/JJ-8/joplin-disable-pdf#readme) | Disable PDF | files | Disables the PDF preview (and export) in Joplin for better performance |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/marc0l92/joplin-plugin-drawio#readme) | Draw.io | integrations | Draw.io (aka Diagram.net) integration for Joplin |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/EliasVincent/joplin-email-note) | Email Note | integrations | Send a note as an email |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/embed-any-search-with-content/14328) | Embed Search | search | Embeds list of links specified by search inside of note |
| -   | embedded tags | tags | embeded tags |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/aman-d-1-n-only/joplin-exports-to-ssg#readme) | Export To SSG | integrations, files | Export a collection of notes to static site generator project. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/benji300/joplin-favorites) | Favorites | note management | Save any notebook, note, to-do, tag, or search as favorite in an extra panel view for quick access. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/persistent-text-folding-in-editor/16183) | Folding in Code Mirror Editor | editor | Allows to fold parts of markdown text displayed in editor |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/tessus/joplin-plugin-get-notebook-id#readme) | Get Notebook ID | developer tools | Copy/Show the ID of a notebook. For developers. A regular user won't need this. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/marc0l92/joplin-plugin-github-theme#readme) | GitHub Theme | theme | Joplin theme with colors of GitHub |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/agerardin/joplin-plugin-knowledge-graph) | Graph | note management | Notes as nodes. Explore your Joplin knowledge graph. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/highlight-matching-plugin/23747) | Highlight Matching | editor | Marks words that match the highlighted selection |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/adarsh-sgh/homenote#readme) | Home Note | note management | Plugin to open a choosen note each time joplin starts. It is like homepages on browsers. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/JackGruber/joplin-plugin-hotfolder/blob/master/README.md) | Hotfolder | files, note management | Monitors a locale folder and import the files as a new note. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/plugin-inline-tags/14192) | Inline tags | tags | Inline tags plugin |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/herdsothom/joplin-insert-date) | Insert Date | editor | Adds a button to insert the current date (without time) into the editor. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/marc0l92/joplin-plugin-jira-issue#readme) | Jira Issue | integrations | Retrieve Atlassian Jira issues information using their api in order to track the status of them from your Joplin notes. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/chenlijun99/autoanki/packages/joplin-plugin) | Joplin Anki Sync | integrations | Bidirectional sync between Joplin and Anki |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/makaanneo/joplin-plugin-apollo) | Joplin Note Attachement rename | files | Joplin plugin to rename note attachments, |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/Cologler/userlink-joplin) | Joplin User Link Plugin | note management | A Joplin plugin for custom links, like `jump://?query=blabla`. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/plugin-share-note-online-via-hackmd/22048) | joplin-hackmd | integrations | Quick share notes on HackMD |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ThibaultJanBeyer/joplin-sheets) | JSheets | editor | JSheets - use to write spreadsheets like excel in Joplin |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/joplin/plugin-kanban/wiki) | Kanban | productivity | Flexible kanban board plugin for all your tasks |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-life-calendar) | Life Calendar | productivity | Your Life in Week Plugin for Joplin |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/treymo/joplin-link-graph) | Link Graph UI | note management | Visualize the connections between Joplin notes. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ajilderda/joplin-macos-native-theme) | macOS theme | theme | Native looking macOS theme for Joplin. Also works on non-macOS devices. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/S73ph4n) | Make All Links | search, note management | Searches the current note for mentions of other notes, then makes the corresponding links. Just click the new icon in the toolbar. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/oswida/joplin-markdown-calc) | Markdown table calculations | editor | Plugin for automatic calculations of markdown table formulas. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-markdown-table-colorize) | Markdown Table: Colorize | appearance, editor | Add colors to markdown table syntax to help distinguish different columns |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-markdown-table-sortable) | Markdown Table: Sortable | editor | Sort rendered markdown tables on the fly. Can recognize simple cases of number, money, date and sort accordingly. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/CalebJohn/joplin-math-mode#readme) | Math Mode | editor | Turn your notes into a powerful calculator with inline math. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/tessus/joplin-plugin-menu-shortcut-toolbar#readme) | Menu items, Shortcuts, Toolbar icons | -----None----- | Additional menu items, shortcuts, and toolbar icons, which are not part of Joplin core |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-metis) | Metis | productivity | A Simple Task Manager Plugin for Joplin based on Todo.txt Specification |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/fengqiaozhu/joplin_plugin_nlr.git) | NLR | integration | 将joplin在记笔记的同时，变成一个网文小说阅读器,a chinese net literature downloader and reader plugin for joplin. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ylc395/joplin-plugin-note-link-system) | Note Link System | note management | A complete Link System for Joplin. Referrer(aka. backlink), Quick Link, Copy Anchor, Hover to preview, and much more feature |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/laurent22/joplin/tree/dev/packages/plugins/ToggleSidebars) | Note list and sidebar toggle buttons | appearance | Adds buttons to toggle note list and sidebar |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/JackGruber/joplin-plugin-note-overview/blob/master/README.md) | Note overview | note management | A note overview is created based on the defined search and the specified fields. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/Kaid00/joplin-note-statistics#readme) | Note Statistics | note management | Plugin to get note statistics |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/benji300/joplin-note-tabs) | Note Tabs | note management, editor | Allows to open several notes at once in tabs and pin them. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/DanteCoder/JoplinPluginNoteVariables#readme) | Note Variables | -----Not Sure----- | A plugin to create variables that can be accesed through all the notes. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/Mick2nd/Notes-Import-Joplin#readme) | Notes Station Import | integrations | This plugin imports notes from QNAP Notes Station |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ylc395/joplin-plugin-ocr) | OCR | note management | offline OCR auto/manually for images, videos, pdf documents in your Joplin notes |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/cqroot/joplin-outline) | Outline | appearance | Outline (TOC) sidebar plugin for joplin. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ylc395/joplin-plugin-pages-publisher) | Pages Publisher | integrations | Generate static blog website from your picked Joplin notes, and publish to Github Pages, with a few mouse clicks. Support preview in local |
| -   | Paste Special | editor | A plugin used to paste special text into Joplin |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/benji300/joplin-persistent-layout) | Persistent Editor Layout | appearance, editor | Save the editor layout (editor/split view/viewer/rich text) for each note separately with custom tags. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/marc0l92/joplin-plugin-plantUML#readme) | PlantUML2 | integration | Render PlantUML diagram inside your Joplin notes. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-quick-goto) | Quick Goto | note management/productivity | Switch to designated notes using keyboard shortcuts |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/martinkorelic/joplin-plugin-quick-html-tags) | Quick HTML tags | editor | Create quick HTML tags. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/quick-links-plugin/14214) | Quick Links | note management | Create links to other notes |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/benji300/joplin-quick-move) | Quick Move | note management, productivity | Collection of commands to quickly move notes to other notebooks. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/Kaid00/joplin-random-note/blob/master/README.md) | Random note | note management | Opens a random note from your vault |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/mak2002/joplin-remove-images-from-note.git) | Remove Images | editor | Allows to remove images from selected note |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/plugin-repeating-to-dos/16470) | Repeating To-Dos | productivity | A powerful and comprehensive plugin for to-do repetition/recurrence |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/roman-r-m/joplin-plugin-resource-search) | Resource Search Plugin | search | Search in Attached Resources |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/CalebJohn/joplin-rich-markdown#readme) | Rich Markdown | editor, appearance | Helping you ditch the markdown viewer for good. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-rubi-and-furigana) | Rubi and Furigana | editor | Commands that insert appropriate &lt;ruby&gt; and &lt;rt&gt; tags to formatted text to display furigana. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/marcgreen/semantic-joplin/tree/master/similar-notes) | Semantically Similar Notes | note management | Show user which notes are semantically similar to the one they are looking at |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/ambrt/joplin-plugin-send-snippet) | Send snippet to different note | note management | Sends snippet of text to any other note from any notebook |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/JackGruber/joplin-plugin-backup/blob/master/README.md) | Simple Backup | files | Plugin to create manual and automatic backups. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/hieuthi/joplin-plugin-slash-commands) | Slash Commands: Datetime & More | -----None-----/prudctivity maybe | Execute several utilities by typing a command that starts with a slash. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/martinkorelic/joplin-plugin-spoilers) | Spoilers | editor | Create inline spoilers and spoiler blocks with title and extendable body. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/roman-r-m/joplin-plugin-table-formatter) | Table Formatter Plugin | editor | Format the table under the cursor in Markdown editor |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/benlau/joplin-plugin-tag-links) | Tag Links | tags | Insert the link of tags at the footer automatically |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/JackGruber/joplin-plugin-tagging/blob/master/README.md) | Tagging | tags | Plugin to extend the Joplin tagging menu with a coppy all tags and a tagging dialog with more control. (Formerly Copy Tags). |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/forcewake/joplin-tags-generator) | Tags Generator | tags | Plugin for Joplin which can be used to extract keywords from note and assign them as a note's tags |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/joplin/plugin-templates) | Templates | note management | This plugin allows you to create and use templates in Joplin. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/Winbee/joplin-turn-to-chart) | turnToChart | editor, appearance | Transform markdown table into d3js chart |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://github.com/joplin/plugin-victor#readme) | Victor | files | Victor can be used to clear all your data - notes, notebooks, attachments, tags, etc. Convenient to start over. |
| [<img width="16" height="16" src=":/cce3c22a39e445a28b1309ded8eee5d5" class="jop-noMdConv">](https://discourse.joplinapp.org/t/joplin-white-theme) | White Theme | theme | A white theme for joplin |

mak2002 commented:
I had previously suggested "Linking" category but eventually decided to make "Note management" instead. Maybe we can use this one if we further split it.

mak2002 commented:
I know that list is long but if it has been reviewed are categories other than `Note Management` fine?

laurent22 commented:
Thanks for looking into it @mak2002, this plugin list with matching categories is very useful.

The only two things I'd change are these:

- Add a "viewer" category where all the renderers could go. We use "viewer" because it's less technical than "renderer". Plugins like "ABC Sheet Music Plugin", "Admonition markdown extension", "Better Markdown Viewer", etc. would go in that category.

- Remove "note management" because it feels like all these plugins could equally go in the "productivity" category. So basically just keep "productivity". It's possible we're missing a good replacement for "note management" but I can't think of anything at this point.

laurent22 commented on packages/generator-joplin/generators/app/templates/webpack.config.js:
```diff
@@ -32,6 +32,7 @@ const packageJsonPath = `${rootDir}/package.json`;
 const manifest = readManifest(manifestPath);
 const pluginArchiveFilePath = path.resolve(publishDir, `${manifest.id}.jpl`);
 const pluginInfoFilePath = path.resolve(publishDir, `${manifest.id}.json`);
+const allPossibleCategories = ['Appearance', 'Developer tools', 'Productivity', 'Themes', 'Integrations', 'Note management', 'Search', 'Tags', 'Editor', 'Files'];
```
this whole list should be lowercase probably

mak2002 replied:
Done :+1:

laurent22 commented on packages/generator-joplin/generators/app/templates/webpack.config.js:
```diff
@@ -67,10 +68,18 @@ function currentGitInfo() {
 }
 
+function validateCategories(categories) {
+	if (!(categories.length === new Set(categories).size)) throw new Error('Repeated categories are not allowed');
```
It's good you've added a check for this but shouldn't that be `categories.length !== new Set(categories).size)`?

mak2002 replied:
Yes, fixed it.

laurent22 commented on packages/generator-joplin/generators/app/templates/webpack.config.js:
```diff
@@ -67,10 +68,18 @@ function currentGitInfo() {
+function validateCategories(categories) {
+	if (!(categories.length === new Set(categories).size)) throw new Error('Repeated categories are not allowed');
+	categories.forEach(category => {
+		if (!allPossibleCategories.includes(category)) throw new Error(`${category} is not a valid category. Valid Categories are: \n${allPossibleCategories}\n`);
```
As an error message, add `please make sure that the category name is lowercase`

mak2002 replied:
Added.

laurent22 reviewed (commented):
I've added a few more comments to the code. Have you been able to test the plugin generator, to make sure it validates the categories correctly? If you need some help testing this let me know.

mak2002 commented:
Thanks Laurent for reviewing it.

If we are removing `Note management` how about adding "Note Linking" / "Linking" for now as I think there are many plugins that would go under this. It will also make "Productivity" less crowded. I know that I had suggested this before but moved on to note management but still. Or I will just remove the category.

As far as testing goes, I have only tested manually with accented words etc. I could not find specific testing tool for webpack config if there is please let me know. So I guess I will add tests with Jest. I will update the PR as per your comments soon.

laurent22 commented:
Jest testing is going to be hard for this, because it's a script within the template. If for now you could just confirm it works by running the generator that would be enough. Here it's described how to run the generator for development: https://github.com/laurent22/joplin/blob/dev/packages/generator-joplin/README.md#generator-joplin

I see what you mean about Linking but I'm not sure it makes a good category. I guess linking, note management and all that are about PKM (Personal knowledge management), so maybe we could have a "pkm" category (although I wonder if it's too obscure).

mak2002 commented:
I had tested it by replacing the webpack file of another plugin by this one and it ran successfully. I will also test as per instructions in link.  


> so maybe we could have a "pkm" category (although I wonder if it's too obscure).

That would be great I think. It maybe a obscure but there is good chance that someone has heard of this term if they are wanting to use such plugin in my opinion if not its pretty much self-explanatory if we are talking about "Personal knowledge management" itself as a category. or maybe we can find some other term for this.

mak2002 commented on packages/generator-joplin/generators/app/templates/webpack.config.js:
```diff
@@ -67,10 +68,18 @@ function currentGitInfo() {
 	const content = fs.readFileSync(manifestPath, 'utf8');
 	const output = JSON.parse(content);
 	if (!output.id) throw new Error(`Manifest plugin ID is not set in ${manifestPath}`);
+	output.categories && validateCategories(output.categories);
```
added this so that it only call it when categories are present

laurent22 replied:
that's a good point but I think you should do this check in validateCategories ("categories" being undefined is valid) otherwise you're doing validation in two places.

mak2002 replied:
Done :+1:

mak2002 commented on packages/generator-joplin/generators/app/templates/webpack.config.js:
```diff
@@ -67,10 +68,18 @@ function currentGitInfo() {
+function validateCategories(categories) {
+		if ((categories.length !== new Set(categories).size)) throw new Error('Repeated categories are not allowed');
+		categories.forEach(category => {
+			if (!allPossibleCategories.includes(category)) throw new Error(`${category} is not a valid category. Please make sure that the category name is lowercase. Valid Categories are: \n${allPossibleCategories}\n`);
```
added error message suggesting lowercase names

mak2002 commented on packages/generator-joplin/generators/app/templates/webpack.config.js:
```diff
@@ -29,6 +29,7 @@ const userConfig = Object.assign({}, {
 
 const manifestPath = `${srcDir}/manifest.json`;
 const packageJsonPath = `${rootDir}/package.json`;
+const allPossibleCategories = ['appearance', 'developer tools', 'productivity', 'themes', 'integrations', 'viewer', 'search', 'tags', 'editor', 'files'];
```
added 'viewer' category and converted all to lowercase

mak2002 commented:
Sorry for many commits had to deal with merge conflicts for the first time. This won't happen again. I added comments to make it easier to review. There are some indentation issues, but I think I should fix them after this gets reviewed to avoid confusion.

As suggested, I tested it as described in this [link](https://github.com/laurent22/joplin/blob/dev/packages/generator-joplin/README.md#generator-joplin), and it ran successfully. I tested it with accented words, empty string, wrong and repeated category.

laurent22 commented:
Thanks for the update but please run `yarn run linter ./` from the root to fix the linter rules.

laurent22 commented:
Also I'm wondering, is the linter not running automatically when you commit?

mak2002 commented:
Linter is running automatically but I think I commited using VS Code's built in git tool so that may have skipped the linting.

mak2002 commented:
I have fixed the linting issues. PR is ready to be reviewed. 

> so maybe we could have a "pkm" category (although I wonder if it's too obscure).

I had a question, are we going to add "Personal knowledge management" category now or later in future?

mak2002 commented:
Thanks, Laurent and everyone else, for guiding me through this long PR. Forgot to remove the older check will update asap.

Edit: I removed the old check. Hopefully, it is ready to be merged

laurent22 commented:
And actually we also need to update the doc otherwise developers won't know what are the possible categories.

The doc to be updated is at `readme/api/references/plugin_manifest.md`. I would suggest adding an entry to the existing "Plugin manifest" section, and also create a new table with the name of each category and a short description.

mak2002 commented:
Should I make changes in this PR itself or create a new one to update the doc?

laurent22 commented:
It should be part of the same pull request, since one doesn't go without the other.

mak2002 commented:
I have updated the doc with new table.

Daeraxa commented:
I think the descriptions need a bit of work - lots of repetition going on ("specifies a plugin...") and some grammatical errors.

mak2002 commented:
No problem. I will update it.

mak2002 commented:
I have removed the repeating part. Let me know what do you think.

laurent22 commented:
Ok let's merge, thanks for creating this list of categories @mak2002!
